### PR TITLE
issue-54-sp-api

### PR DESCRIPTION
**Pull request by stephen-pl**

### CODE REVIEW

**Comment by stephen-pl** [comment]
Current dependencies on/for this PR:
* main
  * **PR #8** <a href="https://app.graphite.dev/github/pr/data-preservation-programs/spade-tenant/8?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #10** <a href="https://app.graphite.dev/github/pr/data-preservation-programs/spade-tenant/10?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #11** <a href="https://app.graphite.dev/github/pr/data-preservation-programs/spade-tenant/11?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #12** <a href="https://app.graphite.dev/github/pr/data-preservation-programs/spade-tenant/12?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/data-preservation-programs/spade-tenant/12?utm_source=stack-comment).

File: api/v1/sp.go
Comment (stephen-pl) [inline api/v1/sp.go:37]: Also need to go over this model with you @jcace

File: api/v1/sp.go
Comment (jcace) [inline api/v1/sp.go:30]: We don't want to filter on a list of spid's. More useful would be to provide a method for filtering on state. Also this needs to be added to the Godoc above so the swagger docs generate correctly as we didn't define that behavior yet.

i.e, this allow them to return only SPs in `active` or `pending` state, for instance. 

However, it's fine to just return all SPs for the Tenant for now to keep it simple. YOur call

File: api/v1/sp.go
Comment (jcace) [inline api/v1/sp.go:39]: don't use the `Debug()` method anywhere 

There is a way of globally enabling debug output which I like better, that way an env var/flag can turn it on/off without changing the code.  See that in `db/main.go` `ln 19` - this thing should have `logger.Defauylt.LogMode(logger.Debug)` passed in if and only if the user specifies `debug = true`

File: api/v1/sp.go
Comment (jcace) [inline api/v1/sp.go:54]: For any of those methods where an array of `spids` is passed in, i think it's better if we just make the body an array of spids (int) instead of decoding to a JSON object 

(i.e, right now it must look like
```json
[
   { "sp_id": 123 },
   { "sp_id": 456 }
]
```

But much simpler if the body of the request can simply be `[123, 456]`

File: api/v1/sp.go
Comment (jcace) [inline api/v1/sp.go:63]: There needs to be a bit more logic here,

It's only possible to approve a sp that is in `pending` state - if they are `eligible` or `suspended` (or already `active`, i guess), then this should be an error for that one.  So we need to check this

Also instead of string value `"eligible"` you can use the enum value i.e `db.TenantSpStateActive`  (add `	"github.com/data-preservation-programs/spade-tenant/db"` to imports at the top of the file)

Comment (jcace) [reply api/v1/sp.go:63]: the more "gormy" way to do this should be like 

```	a.db.Model(&db.TenantsSPs{}).Where("sp_id in (?)", storageProviderIds.SpId).UpdateColumn("tenant_sp_state", db.TenantSpStateActive)```

btw - can you test and see if that one works?

File: api/v1/sp.go
Comment (jcace) [inline api/v1/sp.go:87]: It's only valid to suspend an SP thats in `active` state, other states are an error

Comment (jcace) [reply api/v1/sp.go:87]: also use enum value ` db.TenantSpStateSuspended` please

File: api/v1/sp.go
Comment (jcace) [inline api/v1/sp.go:110]: Only valid to un-suspend an SP in the suspended state

it should to go ` db.TenantSpStateActive` after yes

**Review by jcace — CHANGES_REQUESTED** [review]
see comments. almost there!

File: api/v1/sp.go
Comment (stephen-pl) [reply api/v1/sp.go:30]: Ill do all of them for now

File: api/v1/sp.go
Comment (jcace) [inline api/v1/sp.go:94]: I like the idea of abstracting this into a function, but I think the exact implementation could be better.

The problem is conceptually this thing should be a "db" operation, not an "api" operation - it doesn't _really_  make sense to be a receiver function on `apiV1`, but should be coupled to the database itself. I think the cleanest way to do this is actually to add a Gorm `BeforeSave`[ lifecycle hook ](https://gorm.io/docs/hooks.html#Object-Life-Cycle) to do the validation - so you can bake this logic into the `db` package i.e under `func (*TenantSPs) beforeSave()`

Once you do this, you don't need this `updateTenantSpState` function - you can make it similar to the other API handlers where you just do a DB operation, and then have the error returned if you try to do something invalid

Comment (jcace) [reply api/v1/sp.go:94]: (I haven't actually used Gorm lifecycle hooks before, just read about them. Interested to see exactly how they work)

Comment (stephen-pl) [reply api/v1/sp.go:94]: when you get back from break we can decided between this and https://github.com/data-preservation-programs/spade-tenant/pull/19

**Pull request closed without merging**
